tracing-opentelemetry/src/layer.rs

```diff
@@ -556,12 +556,14 @@ where
 
     // SAFETY: this is safe because the `WithContext` function pointer is valid
     // for the lifetime of `&self`.
-    unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
+    unsafe fn downcast_raw(&self, id: TypeId) -> Option<NonNull<()>> {
         match id {
-            id if id == TypeId::of::<Self>() => Some(self as *const _ as *const ()),
-            id if id == TypeId::of::<WithContext>() => {
-                Some(&self.get_context as *const _ as *const ())
+            id if id == TypeId::of::<Self>() => {
+                Some(NonNull::new_unchecked(self as *const _ as *mut ()))
             }
+            id if id == TypeId::of::<WithContext>() => Some(NonNull::new_unchecked(
+                &self.get_context as *const _ as *mut (),
+            )),
```

Comment:
nit/TIOLI: we could use `NonNull::from` here:
```suggestion
            id if id == TypeId::of::<Self>() => Some(NonNull::from(self).cast::<()>()),
            id if id == TypeId::of::<WithContext>() => 
                Some(NonNull::from(&self.get_context).cast::<()>()),
```